### PR TITLE
test/tls: Add overaligned_bss_var test

### PR DESCRIPTION
**Pull request by jnohlgard**
The initial TLS template image is placed in flash and a separate area is allocated for the TLS variables for the initial thread. crt0 calls `_init_tls` to set up the initial TLS area. The calculated size of the combined TLS area (`.tdata` + `.tbss`) needs to account for different alignment requirements for `.tbss` and `.tdata`, this is done by aligning the size of `.tdata` to the alignment of `.tbss`.
Add a function `_tls_align()` to get the alignment requirement for the TLS area.
This change also eliminates the need for `ALIGN_WITH_INPUT` on the `.tdata` section in the linker script because it is no longer meant to piggyback on the `.data` initialization code in crt0.

Untested on a real system. @keith-packard do you think you could try the test/tls program on some real boards and see if this is correct?

### CODE REVIEW

File: test/tls.c [hunk 1]
```diff
@@ -44,17 +44,19 @@
 #define DATA_VAL2 0x0a0a0a0a
 
 NEWLIB_THREAD_LOCAL volatile int data_var = DATA_VAL;
-NEWLIB_THREAD_LOCAL volatile int bss_var;
 _Alignas(128) NEWLIB_THREAD_LOCAL volatile int overaligned_data_var = DATA_VAL2;
+NEWLIB_THREAD_LOCAL volatile int bss_var;
+_Alignas(128) NEWLIB_THREAD_LOCAL volatile int overaligned_bss_var;
```
Comment (arichardson) [inline test/tls.c:49]: Can you use a different larger alignment (e.g. 256) to make this test more corner cases?

**Comment by keith-packard** [comment]
I guess I just don't understand the motivation here. It seems that the underlying problem is that you can't use `ALIGN_WITH_INPUT` in lld to align the tls data segment in memory correctly, so you can't use the "clever hack"/"over-optimizing kludge" of inserting the TLS data between .data and .bss so that it gets initialized without any additional code. 

To me, the correct place to fix this is to bring lld up to functional parity with the GNU linker, instead of pushing work to support its limitations out to other projects. Personally, I'm happy having the sample linker script and startup code require use of the GNU linker; that's the only one I've ever managed to make work across a huge variety of embedded applications.

Finally, I'll note that having the sample linker script allocate static space for the first TLS area is already a bit sketchy -- usually that space is allocated in a per-task area during task initialization. As such, having the startup code not explicitly initialize those areas makes it work for that mode too, only the linker script need be modified. Yes, the TLS pointer will get set incorrectly until re-set when that initial task area is allocated an initialized. I'm far happier having a complicated linker script and simple run-time code than the inverse, especially when the simple code ends up supporting more use cases.

**Comment by arichardson** [comment]
The `ALIGN_WITH_INPUT` flag only exists to workaround a backwards-incompatible ld.bfd behaviour change (it aligned sensibly before, but recent bfd requires the flag. ld.lld aligns sensibly, but the flag is not supported since it's a bfd-specific workaround.

**Comment by jnohlgard** [comment]
The problem with the original TLS initialization is that it is brittle and makes the linker script unnecessarily rigid because the .tdata section must be placed right after the .data section, both VMA and LMA, and the .tbss section must be placed immediately after .tdata, but with the correct alignment padding.

I didn't test it to verify but it looks like the original linker script may lead to broken initialization if the alignment requirement of .tbss is greater than the alignment requirement of .tdata, and .tdata happens to be not aligned on a .tbss alignment boundary (but still correctly aligned with regards to its own alignment requirement).

By separating the definition of the .tdata TLS "template image" block from the runtime TLS space used for the initial thread we can align that initial thread TLS space on a correct boundary for both .tdata and .tbss and make sure that the whole TLS space has enough space for everything.

Calling `_init_tls` from crt0 instead of implicitly riding on .data and .bss initialization makes the code more obvious and it also acts as an example on how to do the initialization for anyone wanting to enable TLS in their own OS or threading library.

File: test/tls.c (hunk 1 above)
Comment (jnohlgard) [reply test/tls.c]: Like this? afe378e174a9467a029fb906073cd3fe5575c1be

**Comment by keith-packard** [comment]
> The problem with the original TLS initialization is that it is brittle and makes the linker script unnecessarily rigid because the .tdata section must be placed right after the .data section, both VMA and LMA, and the .tbss section must be placed immediately after .tdata, but with the correct alignment padding.

I'm pretty sure that `.tdata` *must* immediately precede `.tbss` in the linker script so that the TLS offsets are allocated correctly. Otherwise, the TLS block might have huge gaps between those two and require significant overallocation (along with potential issues if the offsets grow larger than the architecture supports). But, I might not understand entirely how the linker assigns addresses to TLS variables (it seems like a bit of a kludge to me).

> I didn't test it to verify but it looks like the original linker script may lead to broken initialization if the alignment requirement of .tbss is greater than the alignment requirement of .tdata, and .tdata happens to be not aligned on a .tbss alignment boundary (but still correctly aligned with regards to its own alignment requirement). 

Yup, I added these tests and it's pretty broken. The linker *does* appear to fix things so that the entire TLS block is aligned to the largest alignment value for all TLS variables. So, if .tdata requires 128-byte alignment while .tbss requires 256, then the .tdata segment is rounded up to a 256-byte multiple in size. This seems reasonable enough as it would otherwise be difficult to communicate the TLS block alignment requirements to the application for new TLS blocks.

Check out PR #412 for the fixes (and additional testing). I think that's almost entirely relevant even if you rewrite the linker script to change how the various TLS blocks are specified.

> In this case, the length of the padding between the end of .tdata and the beginning of .tbss will be shorter than what would be necessary if .tdata was aligned on the same alignment as required by .tbss.

Yeah, and _init_tls actually needs to know the real offset of .tbss from the start of .tdata and not simply assume that .tbss starts where .tdata ends. That bug is independent of how the linker script is written.

> By separating the definition of the .tdata TLS "template image" block from the runtime TLS space used for the initial thread we can align that initial thread TLS space on a correct boundary for both .tdata and .tbss and make sure that the whole TLS space has enough space for everything.

That's not necessary -- simply placing the .tdata and .tbss blocks adjacent in the linker script and computing the distance from ADDR(.tdata) to ADDR(.tbss) means that _init_tls can correctly initialize both pieces, even if not contiguous in memory.

> Calling `_init_tls` from crt0 instead of implicitly riding on .data and .bss initialization makes the code more obvious and it also acts as an example on how to do the initialization for anyone wanting to enable TLS in their own OS or threading library.

This relies on there being an initial TLS block being allocated in the linker script, which needn't be the case for applications which explicitly allocate all TLS blocks on a per-task basis.

I have to thank you for discovering all of these TLS alignment issues; I frankly haven't really been concerned about them as no TLS variables in picolibc require specific alignment values. But, making TLS support generally available to applications is pretty powerful.

File: test/tls.c (hunk 1 above)
Comment (keith-packard) [reply test/tls.c]: I had to laugh when I saw this change -- I made exactly the same addition to the tests when you figured out that there was probably a bug here. Check out pr #412 as that has this fix, along with some additional testing (which caught the problem in _init_tls()).

**Comment by keith-packard** [comment]
In looking at the results of linking with lld vs gnu ld, it's pretty clear that they have very different opinions on how to lay out a TLS segment using the linker script. 

part of a gnu ld .map file from test/tls for thumbv7m:
```
.tdata          0x0000000020000100       0x84 load address 0x00000000000036bc
 *(.tdata .tdata.* .gnu.linkonce.td.*)
 .tdata         0x0000000020000100       0x84 test/tls.p/tls.c.o
                0x0000000020000100                data_var
                0x0000000020000180                overaligned_data_var
...
.tbss           0x0000000020000200      0x188
 *(.tbss .tbss.* .gnu.linkonce.tb.*)
 .tbss          0x0000000020000200      0x104 test/tls.p/tls.c.o
                0x0000000020000200                bss_var
                0x0000000020000300                overaligned_bss_var
...
.tbss_space     0x0000000020000200      0x188
...
.bss            0x0000000020000388      0x730
```

And, a similar part of the .map file from lld (although, I had to remove `. = ADDR(.tbss)` because that caused the linker to fail)
```
20000080     3664       84   128 .tdata
20000080     3664       84   128         test/tls.p/tls.c.o:(.tdata)
       0 ffffffffe00035e4        4     1                 data_var
      80 ffffffffe0003664        4     1                 overaligned_data_var
...
20000200 20000200      188   256 .tbss
20000200 20000200      104   256         test/tls.p/tls.c.o:(.tbss)
     180      180        4     1                 bss_var
     280      280        4     1                 overaligned_bss_var
...
20000388 20000388      188     1 .tbss_space
20000388 20000388      188     1         . = . + SIZEOF ( .tbss )
20000510 20000510        0     1 __tbss_space_start = ADDR ( .tbss_space )
20000510 20000510        0     1 __tbss_space_size = SIZEOF ( .tbss_space )
20000510 20000510      730     4 .bss
```

A couple of obvious differences:

 1. gnu ld does not allocate space in the output file for the .tbss segment while lld appears to. gnu ld has special hacks for TLS that bypass space allocation for this segment because it (normally) doesn't get pre-allocated by the linker and is instead allocated at runtime on a per-thread basis. I assume this works correctly for Linux applications linked with lld?
 2. lld doesn't appear to align the .tbss values correctly in the TLS segment -- `overaligned_bss_var` ends up with an offset of 0x280, which isn't aligned on a 256-byte boundary from the start of the TLS segment. That leaves the static TLS segment working correctly while any on-the-fly TLS segments will be mis-aligned.

It's pretty clear that we cannot share a single linker script for these two linkers; with the existing script, lld mis-allocates the output while also incorrectly assigning addresses for some variables.

**Comment by arichardson** [comment]
CC @MaskRay (ld.lld maintainer) regarding ld.lld behaviour differences and the underaligned .tbss variable. Looking at 
```
thread_local int tdata_var = 123;
alignas(64) thread_local int tdata_var_aligned = 123;
thread_local int tbss_var;
alignas(256) thread_local int tbss_var_aligned;
```
(https://godbolt.org/z/16jEdjY7c) it seems that clang+ld.lld will align .tdata to 64 and .tbss to 256, but bfd will align both to the maximum. All sections and variables look sensible with both linkers, so it must be something in the picolibc linker script that causes this misaligned variable.

I guess the error you got with `. = ADDR(.tbss);` is that there are two overlapping sections?
```
	.tbss_space (NOLOAD) : {
		. = ADDR(.tbss);
		. = . + SIZEOF(.tbss);
	} >ram AT>ram :ram
```
I also don't quite understand why the following is needed (it duplicates .tbss?) so the linker should have allocated space in the RAM memory region for this section? Looking at the goldbolt link it appears that at least for this linux target bfd does allocate space for .tbss and the linker script is:
```
  /* Thread Local Storage sections  */
  .tdata	  :
   {
     PROVIDE_HIDDEN (__tdata_start = .);
     *(.tdata .tdata.* .gnu.linkonce.td.*)
   }
  .tbss		  : { *(.tbss .tbss.* .gnu.linkonce.tb.*) *(.tcommon) }
```

**Comment by keith-packard** [comment]
> (https://godbolt.org/z/16jEdjY7c) it seems that clang+ld.lld will align .tdata to 64 and .tbss to 256, but bfd will align both to the maximum. All sections and variables look sensible with both linkers, so it must be something in the picolibc linker script that causes this misaligned variable.

Because TLS variables are allocated in a single block in normal operation, shouldn't the linker ensure that concatenating .tbss to .tdata yields correct alignment for variables in .tbss?

> I guess the error you got with `. = ADDR(.tbss);` is that there are two overlapping sections?

Yes, because lld allocated space for .tbss while gnu ld did not.

> ```
> 	.tbss_space (NOLOAD) : {
> 		. = ADDR(.tbss);
> 		. = . + SIZEOF(.tbss);
> 	} >ram AT>ram :ram
> ```
> 
> I also don't quite understand why the following is needed (it duplicates .tbss?) so the linker should have allocated space 
in the RAM memory region for this section?

No, .tbss is not normally allocated in RAM at all -- it's just used to assign offsets within the TLS region. .tdata ends up in memory because it contains the initialized values for the per-thread TLS region.

 Looking at the goldbolt link it appears that at least for this linux target bfd does allocate space for .tbss and the linker script is:
> 
> ```
>   /* Thread Local Storage sections  */
>   .tdata	  :
>    {
>      PROVIDE_HIDDEN (__tdata_start = .);
>      *(.tdata .tdata.* .gnu.linkonce.td.*)
>    }
>   .tbss		  : { *(.tbss .tbss.* .gnu.linkonce.tb.*) *(.tcommon) }
> ```

Yeah, the resulting symbol table looks correct for both linkers. In particular, the TLS offsets match:

lld TLS variables:

```
    37: 0000000000000000     4 TLS     GLOBAL DEFAULT   18 tdata_var
    38: 0000000000000040     4 TLS     GLOBAL DEFAULT   18 tdata_var_aligned
    39: 0000000000000100     4 TLS     GLOBAL DEFAULT   19 tbss_var
    40: 0000000000000200     4 TLS     GLOBAL DEFAULT   19 tbss_var_aligned
```
 
GNU ld TLS variables:

```
    22: 0000000000000000     4 TLS     GLOBAL DEFAULT   17 tdata_var
    24: 0000000000000100     4 TLS     GLOBAL DEFAULT   18 tbss_var
    27: 0000000000000200     4 TLS     GLOBAL DEFAULT   18 tbss_var_aligned
    31: 0000000000000040     4 TLS     GLOBAL DEFAULT   17 tdata_var_aligned
```

It's pretty clear the picolibc linker script confuses lld quite a bit. I suspect we'll need to experiment with creating a lld-compatible variant and then adapting the crt code to match. What I'd like to avoid is needing special `_set_tls` code for the two linkers; that seems quite do-able, given the fixes for `__tbss_offset` that recently landed.

**Comment by MaskRay** [comment]
Thanks for making picolibc more compatible with lld:)

> `. = ADDR(.tbss);`
> `. = . + SIZEOF(.tbss);`

Symbol assignments within an output section in GNU ld are confusing.
I still can't figure out all the rules, but just noted down some behaviors in https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities#symbol-assignments-within-an-output-section

For portability, I only recommend the following within an output section

* `sym = .;` define a symbol at the current location
* `. += 4;` advance the location counter. This is preferred over `. = . + 4;`

> `__tls_align` introduced by this patch

I think using `p_align(PT_TLS)` may be better.

> gnu ld does not allocate space in the output file for the .tbss segment while lld appears to.

It may be related to this (see https://maskray.me/blog/2022-11-05-lld-musl-powerpc64#stress-test for an example)

Note: when a SHT_NOBITS section is followed by another section, the SHT_NOBITS section behaves as if it occupies the file offset range. This is because ld.lld does not implement a file size optimization.

This size optimization (shared by GNU ld and objcopy) needs a lot of code for something probably not too useful.
I have seen multiple GNU objcopy patches to fix some edge-case bugs in this area... So not implementing the optimization in GNU ld is likely not a bad choice.

> lld doesn't appear to align the .tbss values correctly in the TLS segment ... overaligned_bss_var

Is there a reproduce instruction? You can assume that I currently know nothing about picolibc...

**Comment by keith-packard** [comment]
> Thanks for making picolibc more compatible with lld:)

Well, all we've actually managed to do so far is discover that the sample picolibc linker script is *not* compatible with lld. I guess that's progress (moving from unknown to known failure?).

> 
> > `. = ADDR(.tbss);`
> > `. = . + SIZEOF(.tbss);`
> 
> Symbol assignments within an output section in GNU ld are confusing. I still can't figure out all the rules, but just noted down some behaviors in https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities#symbol-assignments-within-an-output-section

Yes, in this case it was more of a a validation process -- '.' should already be at the location of .tbss as the allocation of addresses within .tbss shouldn't move the location of '.' as variables within .tbss are not actually placed in memory by the linker, only having offsets within the TLS region assigned. It's this assignment which uncovered the incompatibility between the linker script and lld as '.' had been moved during processing of .tbss.

> > `__tls_align` introduced by this patch
> 
> I think using `p_align(PT_TLS)` may be better.

How would we use that?

> It may be related to this (see https://maskray.me/blog/2022-11-05-lld-musl-powerpc64#stress-test for an example)

I wouldn't think so -- the whole TLS processing within gnu ld is full of magic special case code -- it processes the section normally, then goes back and changes the symbol values from the absolute address to the offset from the base of the TLS segment, then it resets the position to the beginning of the .tbss segment before moving to the next command in the script. It has the appearance of an ugly kludge bolted on when TLS support was added to the linker.

> > lld doesn't appear to align the .tbss values correctly in the TLS segment ... overaligned_bss_var
> 
> Is there a reproduce instruction? You can assume that I currently know nothing about picolibc...

Here's a stand-alone test: https://github.com/picolibc/linker-test

**Comment by MaskRay** [comment]
I have analyzed https://github.com/picolibc/linker-test and think lld works as intended.

```
% readelf -Wl test-lld
...
Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x010000 0x10000000 0x10000000 0x00018 0x00018 RWE 0x10000
  LOAD           0x020200 0x20000200 0x20000200 0x00000 0x00a10 RW  0x10000
  LOAD           0x020000 0x20000000 0x10000018 0x00104 0x00104 RW  0x10000
  TLS            0x020080 0x20000080 0x10000098 0x00180 0x00284 RW  0x100
...
% readelf -Ws test-lld | grep tbss_
     8: 00000180     4 TLS     GLOBAL DEFAULT    4 tbss_var
     9: 00000280     4 TLS     GLOBAL DEFAULT    4 tbss_align_var
```

Note that `p_vaddr % p_align != 0`. This is an unusual but valid case.
lld ensures that the variable virtual address in the TLS initialization image aligned, not `st_value`: (0x20000080 + 0x280) % 0x100 = 0 .

I have a write-up here https://maskray.me/blog/2021-02-14-all-about-thread-local-storage#tls-variants and fixed a FreeBSD rtld-elf bug.

**Comment by keith-packard** [comment]
> I have analyzed https://github.com/picolibc/linker-test and think lld works as intended.

The bug I found was that an application which allocates a 0x100-byte aligned TLS block will then have the `tbss_align_var` positioned at an offset of  0x280, which is not in compliance with its stated 0x100 byte alignment. GNU ld aligns the whole TLS block correctly, positioning it so that the `tbss_align_var` has an offset of 0x200, and the static TLS segment starts at 0x20000100 so that the `tbss_align_var` is also correctly aligned there.

In spite of mis-aligning the `tbss_align_var` offset value, lld over-allocates the TLS block, making it 0x284 bytes instead of 0x204 bytes. In addition, lld inserts an unnecessary additional 0x104 bytes between the static location of `tbss_align_var` and `bss_var`.

If the linker script specification is defined by the GNU ld implementation (and I'm not sure what else would be canonical), then lld demonstrates several significant variances from that spec, one of which results in incorrect program operation and another which results in wasted memory.

If, instead, lld has it's own definition of a similar-but-different linker script, then it would be good to refer to that and create a script which generates correct output for this case.

**Comment by MaskRay** [comment]
> > I have analyzed https://github.com/picolibc/linker-test and think lld works as intended.
> 
> The bug I found was that an application which allocates a 0x100-byte aligned TLS block will then have the `tbss_align_var` positioned at an offset of 0x280, which is not in compliance with its stated 0x100 byte alignment. GNU ld aligns the whole TLS block correctly, positioning it so that the `tbss_align_var` has an offset of 0x200, and the static TLS segment starts at 0x20000100 so that the `tbss_align_var` is also correctly aligned there.

No. As mentioned, it's not a bug. There is no requirement that `st_value` is aligned by the variable alignment.
The actual requirement is that the TLS symbol, when copied from the TLS initialization image to a thread block, its address is aligned. This is satisfied by the lld linked executable.

While I can understand that GNU ld overaligns `.tdata` to be at least as large as `.tbss`, I think the behavior is somewhat questionable (but understandable as glibc Variant II ports still have the bug) as the user doesn't request the alignment for `.tdata`.

> In spite of mis-aligning the `tbss_align_var` offset value, lld over-allocates the TLS block, making it 0x284 bytes instead of 0x204 bytes. In addition, lld inserts an unnecessary additional 0x104 bytes between the static location of `tbss_align_var` and `bss_var`.

I disagree with the claim. Since `.tbss` has a larger alignment than `.tdata`, the linker may add padding before `.tdata` to satisfy the alignment requirement.
lld does it in this case so the p_memsz looks like 0x284.

I don't think GNU ld saves space, since there is a padding before `.tdata`

```
% readelf -WS test-ld
There are 13 section headers, starting at offset 0x20530:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .text             PROGBITS        10000000 010000 000018 00  AX  0   0  4
  [ 2] .data             PROGBITS        20000000 020000 000008 00  WA  0   0  4
  [ 3] .tdata            PROGBITS        20000100 020100 000084 00 WAT  0   0 256 #### overaligned, there is padding before this section
  [ 4] .tbss             NOBITS          20000200 020200 000104 00 WAT  0   0 256
  [ 5] .tbss_space       NOBITS          20000200 020200 000104 00  WA  0   0  1
  [ 6] .bss              NOBITS          20000304 020200 000004 00  WA  0   0  4
  [ 7] .stack            NOBITS          20000308 020200 000800 00  WA  0   0  1
  [ 8] .comment          PROGBITS        00000000 020184 00001e 01  MS  0   0  1
  [ 9] .ARM.attributes   ARM_ATTRIBUTES  00000000 0201a2 000039 00      0   0  1
  [10] .symtab           SYMTAB          00000000 0201dc 000210 10     11  16  4
  [11] .strtab           STRTAB          00000000 0203ec 0000de 00      0   0  1
  [12] .shstrtab         STRTAB          00000000 0204ca 000065 00      0   0  1
```

> If the linker script specification is defined by the GNU ld implementation (and I'm not sure what else would be canonical), then lld demonstrates several significant variances from that spec, one of which results in incorrect program operation and another which results in wasted memory.
>
> If, instead, lld has it's own definition of a similar-but-different linker script, then it would be good to refer to that and create a script which generates correct output for this case.

In this case I think GNU ld's behavior isn't particular good, so I don't think it is fair to port it to lld.

**Comment by MaskRay** [comment]
Note, we can also construct an `align(.tdata)=128 && align(.tbss)=256` case where GNU ld's overaligning of `.tdata` to 256 will make the TLS initialization image larger than lld's. Simply try `size(.tdata)=128`: lld's output has no padding in TLS while GNU ld's output will have a 128-byte padding.

**Comment by jnohlgard** [comment]
Let's keep in mind that the tdata area in the linker script is only used as the template for the per thread tls area, so the alignment of tdata in the binary/flash is not really meaningful other than it should be aligned in a way that it is easy to memcpy it into the thread tls area when a new thread is created. The threading library or os needs to ensure that the thread's tls area is aligned according to the requirements of the variables.
By calculating the tls area alignment as the max of tdata align and tbss align and a placing padding between tdata and tbss we simplify the runtime handling because we know that all tls variables have correctly aligned offsets from the tls pointer. In particular, the space optimization of placing padding before tdata instead of between tdata and tbss leads to some calculations needed in the runtime code to make sure tbss ends up aligned when doing the tls initialization for a newly created thread.
The padding between tdata and tbss is what is being calculated in this pr when we do `PROVIDE( __tls_size = ALIGN(SIZEOF(.tdata), ALIGNOF(.tbss)) + SIZEOF(.tbss) );`
The memset for tbss in `_init_tls` is extended to also clear the padding bytes in this PR, but that was simply a design decision by me that I wanted to avoid leaking some random freed data that could be left from some other code in a bare metal project.
